Streaming Aggregates for Queries

bplatz opened the pull request:
### Summary

Implements incremental streaming aggregation for queries, reducing memory overhead for aggregate queries over large result sets. Instead of collecting all solutions into groups before computing aggregates, this approach updates aggregate state incrementally as solutions flow through the query pipeline.

### Key Changes

- **Streaming aggregate descriptors** (`src/fluree/db/query/exec/eval.cljc`) - Registry of `{:init :step! :final}` functions for `count`, `count-distinct`, `sum`, `avg`, `min`, `max`
- **Eligibility detection** (`src/fluree/db/query/exec/group.cljc`) - Automatically uses streaming path when query structure permits (no HAVING clause, all selectors are group vars or streamable aggregates)
- **Selector updates** (`src/fluree/db/query/exec/select.cljc`) - `AggregateSelector` and `AsSelector` now carry optional `streaming-agg` descriptor
- **Parse-time detection** (`src/fluree/db/query/fql/parse.cljc`) - Builds streaming descriptors for simple aggregate forms like `(count ?x)`, `(sum ?y)`

### How It Works

Traditional grouping collects all solutions, then applies aggregate functions:
solutions → [collect all] → group → apply agg-fn → results

Streaming aggregation updates state incrementally:
solutions → step!(state, value) per solution → final(state) → results

### Supported Aggregates

| Function | Streaming Support |
|----------|-------------------|
| `count` | ✅ |
| `count` with `*` | ✅ |
| `count-distinct` | ✅ |
| `sum` | ✅ |
| `avg` | ✅ |
| `min` | ✅ |
| `max` | ✅ |

### Limitations

Falls back to traditional grouping when:
- Query has a `HAVING` clause
- Aggregate uses complex expressions (e.g., `(sum (+ ?x ?y))`)
- Non-streamable aggregate functions (e.g., `sample`, `group_concat`)

bplatz committed:
Add index-range method to IFlureeDb protocol and implement in AsyncDB and FlakeDB

bplatz committed:
Implement streaming object aggregates (COUNT, AVG, SUM, MIN, MAX) in FQL

bplatz committed:
Implement streaming aggregates and enhance selector handling for incremental evaluation

bplatz committed:
Merge branch 'main' into perf/streaming-aggregate

bplatz committed:
Merge remote-tracking branch 'origin/main' into perf/streaming-aggregate

bplatz committed:
Enhance streaming aggregation logic to support implicit grouping and improve logging

bplatz committed:
Refactor extract-group-key function to optimize group key extraction in streaming mode

bplatz committed:
Improve nil handling in min and max step functions for streaming aggregates

bplatz committed:
Enhance sum and avg step functions to handle 0 and false values in streaming aggregates

bplatz committed:
Refactor streaming aggregate descriptors for clarity and efficiency; enhance aggregate function handling in select and group processing

bplatz committed:
Refactor streaming aggregation logic to enhance eligibility checks and streamline group updates; improve finalization of aggregate states

zonotope commented on src/fluree/db/query/exec/select.cljc:
```diff
@@ -60,53 +60,73 @@
       (with-meta selector {`format-value select.fql/format-wildcard-selector-value}))))
 
-(defrecord AggregateSelector [agg-fn]
+(defrecord AggregateSelector [agg-fn streaming-agg]
```
I think I would prefer a `StreamingAggregateSelector` record with its own separate `ValueSelector` implementation. Having this opaque map complects multiple behavior traits into the same record.

bplatz replied:
Addressed in 7c57470b9a52c69e9ab2b6f69a0219e4ab4fa64f

zonotope reviewed (commented):
I'd like to start a discussion about the approach and whether or not my suggestions are feasible.

bplatz committed:
Merge remote-tracking branch 'origin/main' into perf/streaming-aggregate

bplatz committed:
Refactor streaming aggregate logic to unify step function naming and enhance selector handling; improve error logging for aggregate selectors

bplatz committed:
Implement streaming aggregators for incremental group aggregation; refactor related functions in eval and group modules

zonotope commented on src/fluree/db/query/exec/aggregate.cljc:
```diff
@@ -0,0 +1,219 @@
+      (inc state)
+      state))
+  (finalize [_ state]
+    (where/->typed-val state)))
```
I think the code would be simpler if the aggregator records encapsulated their state. Then you could use the aggregators themselves to keep track of the running tally. I also would remove the `initialize` method in favor of a constructor. Here's how I'd change the `CountAggregator`:

```suggestion
(defrecord CountAggregator [tally]
  StreamingAggregator
  (step [this tv]
    (if (some-> tv :value some?)
      (update this :tally inc)))
  (finalize [_]
    (where/->typed-val tally)))

(defn count-aggregator
  []
  (->CountAggregator 0))
```

I'd change the other records similarly.

zonotope commented on src/fluree/db/query/exec/select.cljc:
```diff
@@ -63,50 +63,82 @@
-   (cond-> (->AggregateSelector agg-function)
+   (->AggregateSelector agg-function))
+  ([agg-function streaming-agg]
+   (if streaming-agg
```
I think this conditional should be lifted to the caller of this constructor function, and this function should be replaced with two different functions. In general I think conditionals should be evaluated as soon as possible. This simplifies the code by making clear execution branches instead of putting everything together and transmitting conditionals without context. If there's shared functionality between the two branches, then that could still be implemented with helper functions.

zonotope commented on src/fluree/db/query/exec/group.cljc:
```diff
@@ -95,17 +101,103 @@
+        streaming-aggs (->> selectors
+                            (filter streaming-agg-selector?)
+                            (mapv :streaming-agg))
+        implicit?      (and (empty? group-vars)
```
If this is necessary, then I think it's a bug unless I'm missing something. `select/implicit-grouping?`. If I remember correctly, `select/implicit-grouping?` should only return a truthy value if there are no grouped variables.

zonotope commented on src/fluree/db/query/exec/group.cljc:
```diff
@@ -95,17 +101,103 @@
+                            (mapv :streaming-agg))
+        implicit?      (and (empty? group-vars)
+                            (some select/implicit-grouping? selectors))
+        streaming?     (streaming-eligible? having streaming-aggs implicit?
```
I don't think this needs to be a separate binding. I think `(if (streaming-eligible? having ...)` is just as if not more readable.

bplatz committed:
refactor, pr feedback

bplatz commented:
@zonotope ready for re-review:

Changes Made
1. aggregate.cljc - Protocol conversions
Converted ->offset-date-time multimethod to OffsetDateTimeConverter protocol for better JVM dispatch performance
Converted ->offset-time multimethod to OffsetTimeConverter protocol
2. aggregate.cljc - Aggregator state encapsulation
Refactored StreamingAggregator protocol to encapsulate state in records
Removed initialize method; aggregators now start with initial state via constructor functions
Changed step signature from (step this state tv) to (step this tv) - returns new aggregator
Renamed finalize to complete (to avoid conflict with Java's Object.finalize())
Updated all aggregator records: CountAggregator, CountStarAggregator, CountDistinctAggregator, SumAggregator, AvgAggregator, MinAggregator, MaxAggregator
streaming-aggregators map now stores constructor functions
3. group.cljc
Added imports for AsSelector, StreamingAggregateSelector, VariableSelector to avoid long fully-qualified names
Updated streaming-agg-selector? and streaming-eligible? to use short names
Updated docstring for update-streaming-groups
Updated update-streaming-groups and finalize-streaming-groups to work with new aggregator interface
Inlined the streaming? binding into the if condition
4. select.cljc
Split aggregate-selector into two separate functions:
aggregate-selector - for non-streaming aggregates
streaming-aggregate-selector - for streaming aggregates
Removed conditional logic from the function (lifted to caller)
5. parse.cljc
Updated parse-select-aggregate to explicitly choose between aggregate-selector and streaming-aggregate-selector

bplatz commented:
Note on ci/cd errors, this is fixed with https://github.com/fluree/db/pull/1190 -- we use 'latest' version of clj-kondo and I think a new version is flagging these.

bplatz committed:
Merge branch 'main' into perf/streaming-aggregate

zonotope commented on src/fluree/db/query/exec/aggregate.cljc:
```diff
@@ -0,0 +1,256 @@
+      (update this :tally inc)
+      this))
+  (complete [this]
+    (where/->typed-val (:tally this))))
```
Small nit, but tally is already bound, so you can just do this:

```suggestion
    (where/->typed-val tally)))
```

zonotope commented on src/fluree/db/query/exec/aggregate.cljc:
```diff
@@ -0,0 +1,256 @@
+                        :status 400
+                        :error  :db/invalid-fn-call})))))
+
+(defn compare*
```
I think this could have a slightly more descriptive name like `compare-values`.

zonotope commented on src/fluree/db/query/exec/aggregate.cljc:
```diff
@@ -0,0 +1,256 @@
+         :cljs (compare val-a val-b))
+
+      ;; same types compare
+      (= dt-a dt-b)
```
this might not matter that much in practice, but I feel like this case should come first since equality is the least expensive check we do in this function.

zonotope commented on src/fluree/db/query/exec/aggregate.cljc:
```diff
@@ -0,0 +1,256 @@
+  (step [this tv]
+    (let [cur (:current this)]
+      (cond
+        (nil? (some-> tv :value)) this
```
The `nil?` and `some->` here is redundant. We can just do this:

```suggestion
        (-> tv :value nil?) this
```

zonotope commented on src/fluree/db/query/exec/aggregate.cljc:
```diff
@@ -0,0 +1,256 @@
+  (step [this tv]
+    (let [cur (:current this)]
+      (cond
+        (nil? (some-> tv :value)) this
```
Same here for the redundant `nil?` and `some->`

```suggestion
        (-> tv :value nil?) this
```

zonotope commented on src/fluree/db/query/exec/aggregate.cljc:
```diff
@@ -0,0 +1,256 @@
+        (neg? (compare* tv cur)) (assoc this :current tv)
+        :else this)))
+  (complete [this]
+    (or (:current this) (where/->typed-val nil))))
```
This `or` is redundant. `current` should never be false, so the first branch of the `or` is falsey only if `current` is `nil`, so the second branch shouldn't be necessary. This should work:

```suggestion
    (where/->typed-val current)))
```

zonotope commented on src/fluree/db/query/exec/aggregate.cljc:
```diff
@@ -0,0 +1,256 @@
+        (pos? (compare* tv cur)) (assoc this :current tv)
+        :else this)))
+  (complete [this]
+    (or (:current this) (where/->typed-val nil))))
```
Same here

```suggestion
    (where/->typed-val current)))
```

zonotope commented on src/fluree/db/query/exec/group.cljc:
```diff
@@ -95,17 +102,102 @@
+              :agg-states     {}})
+        agg-states'
+        (reduce (fn [states {:keys [arg-var result-var aggregator]}]
+                  (let [agg    (get states result-var (aggregator))
```
the spacing here is off

zonotope commented on src/fluree/db/query/exec/group.cljc:
```diff
@@ -95,17 +102,102 @@
+                                         (get arg-var)
+                                         where/mch->typed-val))
+                        agg'   (agg/step agg tv)]
+                    (assoc states result-var agg')))
```
I think you can get rid of both the `agg` and `agg'` bindings above and just do

```clojure
(update states result-var (fnil agg/step (aggregator)) tv)
```

you can take or leave that. I prefer it since it's less code, but the `fnil` makes it a little more complex so i get if you prefer how it is now.

zonotope reviewed (approved):
🎒
